pkg/gameservers/health.go

```diff
@@ -125,8 +125,11 @@ func (hc *HealthController) unschedulableWithNoFreePorts(pod *corev1.Pod) bool {
 func (hc *HealthController) failedContainer(pod *corev1.Pod) bool {
 	container := pod.Annotations[agonesv1.GameServerContainerAnnotation]
 	for _, cs := range pod.Status.ContainerStatuses {
-		if cs.Name == container && cs.State.Terminated != nil {
-			return true
+		if cs.Name == container {
+			// sometimes on a restart, these cs.State can be running and the last state will be merged
```

Comment:
question - was this a bug before where there was a race between checking the health and the container restarting?
question2 - it seems like this will always return true if the container has *ever* failed, which is a change from before (where it was transient but could flip back to false after a restart). Is that the correct behavior?

Reply:
These are good questions:

question1 - So the bug was a couple of things:
1) The issue outlined in #956 , but also, that could potentially not happen (i.e. a race condition could mean on restart, it would remain in a Ready state) -- so it was inconsistent. At least now, it consistently moves to an unhealthy state on crash.

2) This is essentially true. If the pod has ever failed (in the last 2 iterations, because that is all that is tracked), and the pod gets updated (say the sidecar crashes), we will move to unhealthy - but at least consistently so.

The next PR I'm going to do, is actually solve #956. My current thinking is - when a GameServer goes to Ready, we put an annotation on the pod with the current `Running` gameserver container ID. Then we always know what not to look at before - i.e. if there was a status.State of Running, and it matched our annotation, we would know not to look at the LastTerminatedState, since that came before.

If that annotation is not there, we don't ever move to Unhealthy, and allow restarts to occur as per normal K8s.

There might be some other edge cases in there -- but I think that gets us to what is the currently documented health checking strategy.

TL;DR - this is an interim step.

Reply:
I just realised that 1.0 RC comes out Tuesday. Happy for this to be merged post 1.0 stable release. It means we have #956 outstanding for 1.0, but I don't think it's a hard blocker for 1.0.

I wouldn't want this to be included in 1.0 without the follow up PR to fix #956 - I've started work on it, but probably safer to wait.